### PR TITLE
Reproschema to fhir json script

### PR DESCRIPTION
**Pull request by Evan8456**
- Updated old script to for reproschema to fsh
- Added a new script that converts reproschema to fhir json
- Added GAD7 and demographics questionnaire to test
- Script generates the questionnaire json and a json file containing a list of valuesets and codesystems

### CODE REVIEW

File: input/reproSchemaScript/reprotofhirjson.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        ]
+
+        valueset["description"] = id
+        valueset["compose"] = dict()
+        
+        valueset["compose"]["include"] = [{"system": f"{CODESYSTEM_URI}{id}"}]
+
+        fhir_valuesets.append(valueset)
+
+    curr_question["answerValueSet"] = str(VALUESET_URI) + id
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:135]: try to consistently use f strings, will make it easier to read

File: input/reproSchemaScript/reprotofhirjson.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        # add options
+        if "responseOptions" in question_json:
+            curr_question = add_options(curr_question, question_json, questionnaire, i["variableName"],  mode)
+        group["item"].append(curr_question)
+
+    fhir_questionnaire["item"].append(group)
+  
+
+if __name__ == '__main__':
+    convert_to_fsh(sys.argv[1], sys.argv[2] )
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:264]: use argparse for the input arguments: https://docs.python.org/3/library/argparse.html

File: input/reproSchemaScript/reprotofhirjson.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    questionnaire_name = sys.argv[1]
+    q = json.dumps(fhir_questionnaire)
+
+    f = open(f"{questionnaire_name}.json", "w+")
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:268]: doing `with open(...) as f:` is more idiomatic

File: input/reproSchemaScript/reprotofhirjson.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    f.write(str(q))
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:269]: why do you need to cast q as a string here?

File: input/reproSchemaScript/reprotofsh-old.py
Comment (alistairewj) [inline input/reproSchemaScript/reprotofsh-old.py:1]: you can just delete the file. we have the git history to restore/go back if we want

File: input/reproSchemaScript/reprotofhirjson.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+from dotenv import load_dotenv
+
+load_dotenv()
+CODESYSTEM_URI = os.getenv('CODESYSTEM_URI')
+VALUESET_URI = os.getenv('VALUESET_URI')
+QUESTIONNAIRE_URI = os.getenv('QUESTIONNAIRE_URI')
+
+codeSystem_dict = dict()
+
+fhir_questionnaire = dict()
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:22]: having globals like this is a bad habit, because it makes the code less modular and it's hard to follow. instead define this within the function and return it.

File: input/reproSchemaScript/reprotofhirjson.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    ]
+
+    codeSystem["description"] = id
+    codeSystem["caseSensitive"] = True
+    codeSystem["content"] = "complete"
+    codeSystem["count"] = len(options_json["choices"])
+    codeSystem["concept"] = []
+    
+    options = []
+    # loops through the options and appends them to the codesystem
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:68]: In general this is the type of comment you want to avoid making, because it's *too* literal.

It tells me that you loop through the options and then append them. Then I read the code and the code is `for j in options_json ...` (a loop) followed by `options.append(choice)` (an append). So in fact the comment hasn't given me much added info over the code itself. Instead try to write a comment that explains *why* you are doing this. My understanding is you are collecting each choice in the reproSchema input and converting them into a code/display to be added to the code system in FHIR. So the ideal comment would say: (1) what does reproSchema store in "choices" (e.g. what is their data structure), and (2) how is it equivalently stored in FHIR (what's your target structure), and (3) how you are going to go from A->B

File: input/reproSchemaScript/reprotofhirjson.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    for j in options_json["choices"]:
+        codeSystem_option = dict()
+        if "schema:name" in j and j["schema:name"] != "":
+            choice = j["schema:name"]
+        else:
+            choice = j["schema:value"]
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:74]: is `schema:value` guaranteed to exist?

File: input/reproSchemaScript/reprotofhirjson.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        
+        if choice and not isinstance(choice, int) and "en" in choice and isinstance(choice, dict):
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:76]: does this mean we are ignoring internationalization currently? should note this down somewhere

File: input/reproSchemaScript/reprotofhirjson.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    if not exists:
+
+        valueset = dict()
+        valueset["resourceType"] = "ValueSet"
+        valueset["id"] = id
+        valueset["text"] = {
+        "status" : "generated",
+        "div" : "<div xmlns=\"http://www.w3.org/1999/xhtml\">Placeholder</div>"
+        }
+        valueset["url"] = str(VALUESET_URI) + id
```
Comment (alistairewj) [inline input/reproSchemaScript/reprotofhirjson.py:108]: I don't think you need `str()` here as os.getenv returns a string

**Review by alistairewj — CHANGES_REQUESTED** [review]
just some very quick comments before we move this into its own repo

is there any way you can add the reproschema stuff as a submodule instead of copy pasting the files into here?

**Comment by Evan8456** [comment]
> just some very quick comments before we move this into its own repo
> 
> is there any way you can add the reproschema stuff as a submodule instead of copy pasting the files into here?

I originally added the reproschema stuff for easy access. In practice, all you need to pass in is path to the reproschema stuff. I could either do that via command line parameters(user just passes in the path to the reproschema they want to convert) or I can add it as a environment variable where you put the absolute path to the reproschema.

**Comment by alistairewj** [comment]
> > just some very quick comments before we move this into its own repo
> > is there any way you can add the reproschema stuff as a submodule instead of copy pasting the files into here?
> 
> I originally added the reproschema stuff for easy access. In practice, all you need to pass in is path to the reproschema stuff. I could either do that via command line parameters(user just passes in the path to the reproschema they want to convert) or I can add it as an environment variable where you put the absolute path to the reproschema.

having it as an input argument makes sense! and then in the README, you can provide a git clone command to get the commit of data you pulled for running the script

**Review by alistairewj — APPROVED** [review]
